jetbrains-core/src/software/aws/toolkits/jetbrains/core/credentials/ToolkitCredentialProcessProvider.kt

```diff
@@ -0,0 +1,101 @@
+import com.intellij.openapi.util.registry.Registry
+import com.intellij.util.execution.ParametersListUtil
+import software.amazon.awssdk.auth.credentials.AwsBasicCredentials
+import software.amazon.awssdk.auth.credentials.AwsCredentials
+import software.amazon.awssdk.auth.credentials.AwsCredentialsProvider
+import software.amazon.awssdk.auth.credentials.AwsSessionCredentials
+import software.amazon.awssdk.utils.cache.CachedSupplier
+import software.amazon.awssdk.utils.cache.RefreshResult
+import software.aws.toolkits.core.utils.debug
+import software.aws.toolkits.core.utils.error
+import software.aws.toolkits.core.utils.getLogger
+import software.aws.toolkits.resources.message
+import java.time.Instant
+
+class ToolkitCredentialProcessProvider internal constructor(
+    private val command: String,
+    private val parser: CredentialProcessOutputParser
+) : AwsCredentialsProvider {
+    constructor(command: String) : this(command, DefaultCredentialProcessOutputParser)
+
+    private val entrypoint by lazy {
+        ParametersListUtil.parse(command).first()
+    }
+    private val cmd by lazy {
+        if (SystemInfo.isWindows) {
+            GeneralCommandLine("cmd", "/C", command)
+        } else {
+            GeneralCommandLine("sh", "-c", command)
+        }
+    }
+    private val processCredentialCache = CachedSupplier.builder { refresh() }.build()
+
+    override fun resolveCredentials(): AwsCredentials = processCredentialCache.get()
+
+    private fun refresh(): RefreshResult<AwsCredentials> {
+        val timeout = Registry.intValue("aws.credentialProcess.timeout", DEFAULT_TIMEOUT)
+        val output = ExecUtil.execAndGetOutput(cmd, timeout)
+
+        if (output.isTimeout) {
+            handleException(message("credentials.profile.credential_process.timeout_exception_prefix", entrypoint), output)
+        }
+
+        if (output.exitCode != 0) {
+            handleException(message("credentials.profile.credential_process.execution_exception_prefix", entrypoint), output)
+        }
+
+        val result = try {
+            parser.parse(output.stdout)
+        } catch (e: Exception) {
+            handleException(message("credentials.profile.credential_process.parse_exception_prefix"), output, e)
+        }
+        val credentials = when (val token = result.sessionToken) {
+            null -> AwsBasicCredentials.create(result.accessKeyId, result.secretAccessKey)
+            else -> AwsSessionCredentials.create(result.accessKeyId, result.secretAccessKey, token)
+        }
+        return RefreshResult.builder(credentials).staleTime(result.expiration ?: Instant.MAX).build()
+    }
+
+    private fun handleException(msgPrefix: String, process: ProcessOutput, exception: Exception? = null): Nothing {
+        val errorOutput = process.stderr.takeIf { it.isNotBlank() }
+        val details = errorOutput ?: exception?.message?.takeIf { it.isNotBlank() } ?: exception?.cause?.message?.takeIf { it.isNotBlank() }
```

Comment:
Hmmm...

Should the cause take precedence over the parent exception? Wouldn't that show the more accurate error in something like a JSON parse exception?

Reply:
Also port over the sanitize logic that the SDK has: https://github.com/aws/aws-sdk-java-v2/pull/1166/files#diff-f0e43a150bd11c291e828f1ecdb3ffd1R90